Improvement to Find/FindAsync with Partition Key

1iveowl opened the pull request:
<!--
Please check if the PR fulfills these requirements

- [ ] The code builds and tests pass (verified by our automated build checks)
- [ ] Commit messages follow this format
        Summary of the changes
        - Detail 1
        - Detail 2

        Fixes #bugnumber
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Code meets the expectations our engineering guidelines.
      https://github.com/dotnet/aspnetcore/wiki/Engineering-guidelines

Review the guidelines for CONTRIBUTING.md for more details.
-->

Suggested improvements for `Find`/`FindAsync` for Cosmos, allowing for ReadItem optimization, when the Partition Key is known **and** Cosmos Resource Id is provided **or** Cosmos Resource Id can be generated based on the default generator or a custom generator. 

Please see #17310

1iveowl commented on src/EFCore.Cosmos/ValueGeneration/Internal/IdValueGenerator.cs:
```diff
@@ -47,7 +48,7 @@ protected override object NextValue(EntityEntry entry)
 
             var partitionKey = entityType.GetPartitionKeyPropertyName() ?? CosmosClientWrapper.DefaultPartitionKey;
-            foreach (var property in pk.Properties)
+            foreach (var property in pk.Properties.Where(p => p.Name != CosmosConstant.ResourceIdIdentifier))
```
The resource `id` itself should be ignored here.

1iveowl commented:
> The fix should not change `ParameterExtractingExpressionVisitor` in anyway. That is not what it meant for. Wait for @AndriySvyryd to provide his thoughts on how to fix.

Thank you @smitpatel. Changing it shouldn't be a big issue. The hard part was not writing the code, rather it was figuring out how it all works together in the Cosmos provider and then pick the best way implement the suggested improvements, in a manner that best aligns with the existing and with minimal changes.

Regarding `ParameterExtractingExpressionVisitor`, please let me share my thinking behind why I picked this, as that will help you understand my thinking and thus should help you guiding me where you'd rather want me to go.

Technically I haven't changed `ParameterExtractingExpressionVisitor`, I've extended it. I did this exactly because I didn't want to rewrite the logic of `ParameterExtractingExpressionVisitor` by creating a new visitor for traversing `query`. `ParameterExtractingExpressionVisitor` is doing the exact same thing it did before. By extending it, I'm extracting some additional parameters from `query`. Parameters that are needed for the ReadItem optimization. 

Anyhow, I'm sure there are other ways to collect those needed parameters. Would be great if your and/or @AndriySvyryd could point me in the direction of where you'd want me to do the changes.

smitpatel commented:
> I'm sure there are other ways to collect those needed parameters.

What are the parameters you need to collect?

ajcvickers commented:
@1iveowl 

> The hard part was not writing the code, rather it was figuring out how it all works together in the Cosmos provider and then pick the best approach for implementing the suggested improvements, in a manner that best aligns with the existing pattern etc., and with minimal changes.

Believe me, this is how I feel whenever I look at the query code! 😸 Thank you for your efforts here!

1iveowl commented:
> > I'm sure there are other ways to collect those needed parameters.
> 
> What are the parameters you need to collect?

@smitpatel what I need to extract from the query is:
- The partition key name and value
- The name and value pairs of all the primary keys.
- The resource id value, if it is provide

Seems easy enough, right 😊

I also look to see if the query is attempting to fetch only one item - i.e if the ShapedQueryExpression has `ResultCardinality.Single` or `ResultCardinality.SingleOrDefault`.

It all comes together [here](https://github.com/1iveowl/efcore/blob/a75693a649ea7f19f2993045c3970b233a224f93/src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.QueryingEnumerable.cs#L57) where it is decided if the query can use the new `ReadItem` optimization can be used or if the orogonal `FeedIterator` should be used.

smitpatel commented:
So assumptions about first EntityType, setting EntityType/PK is all incorrect thing to put on QueryContext. If you see in translation phase, you will have a query root expression which will tell you which IEntityType it is. From IEntityType you can get IProperties which makes PK. From IEntityType you can also fetch IProperty for "id" and it's value generator. So from query perspective, whole processing should be around compilation time, no changes should be made to ParameterExtractingEV.
The best approach would be look at the expression tree you get inside QueryTranslationPreprocessor for Find query and see how you can identify that pattern and provide an alternate translation for it.
Also from CosmosClientWrapper side, it would be good to keep the method which gets one item using ReadItem separate from FeedIterator. You could using QueryingEnumerable from query side but perhaps generate a different enumerable.

1iveowl commented:
> If you see in translation phase, you will have a query root expression which will tell you which IEntityType it is

What about `model.GetEntityTypes().First()` is this a no-no or?

smitpatel commented:
It's a big no-no.

1iveowl commented:
> The best approach would be look at the expression tree you get inside QueryTranslationPreprocessor

I.e. using the `CosmosQueryTranslationPreprocessor` and the `CosmosQueryMetadataExtractingExpressionVisitor` that I created for the `WithParitionKey` extension?

1iveowl commented:
> If you see in translation phase, you will have a query root expression which will tell you which IEntityType it is

[This query](https://github.com/dotnet/efcore/blob/d862edca59c9e5468e67df63b58dbadb643bb669/src/EFCore.Cosmos/Query/Internal/CosmosQueryTranslationPreprocessor.cs#L25) passed to the `NormalizeQueryableMethodCall`?

smitpatel commented:
Yes, in that method you can inspect the expression tree. (Use `new ExpressionPrinter().Print(query)` to get a good debug output). From we can figure out how do we reach the final query which we want to. Can you also paste debug output here?

1iveowl commented:
> Yes, in that method you can inspect the expression tree. (Use `new ExpressionPrinter().Print(query)` to get a good debug output). From we can figure out how do we reach the final query which we want to. Can you also paste debug output here?

Here's the debug of the query:

> .Call System.Linq.Queryable.FirstOrDefault(
    .Extension<Microsoft.EntityFrameworkCore.Query.QueryRootExpression>,
    '(.Lambda #Lambda1<System.Func`2[Microsoft.EntityFrameworkCore.Cosmos.EndToEndCosmosTest+Customer_WithResourceId,System.Boolean]>))

> .Lambda #Lambda1<System.Func`2[Microsoft.EntityFrameworkCore.Cosmos.EndToEndCosmosTest+Customer_WithResourceId,System.Boolean]>(Microsoft.EntityFrameworkCore.Cosmos.EndToEndCosmosTest+Customer_WithResourceId $e)
{
    .Call Microsoft.EntityFrameworkCore.EF.Property(
        $e,
        "PartitionKey") == $__p_0 && .Call Microsoft.EntityFrameworkCore.EF.Property(
        $e,
        "id") == $__p_1 && .Call Microsoft.EntityFrameworkCore.EF.Property(
        $e,
        "Id") == $__p_2
}

smitpatel commented:
`Microsoft.EntityFrameworkCore.Query.QueryRootExpression` has EntityType property.

smitpatel commented:
In first pass, let's just optimize queries coming from Find where PK properties contain id & partition key both.
Step 1: Identify the query shape and skip the translation pipeline.
Override `CosmosQueryableMethodTranslatingExpressionVisitor.Visit` method. You will get same query expression as above. You pattern match it with following
- MethodCallExpression
- Method is Queryable.FirstOrDefault
- Source is QueryRootExpression - get entityType from QueryRootExpression
- predicate is non-null
- Predicate is binary expression with ExpressionType.AndAlso where binaryExpression.Left is ExpressionType.Equal and binaryExpression.Right is either ExpressionType.AndAlso with recursive structure or Expression.Equal.
- Collect all equalty comparisons from predicate, Left side of all comparison should EF.Property method call.
- Use EntityType from QueryRootExpression and use propertyname from EF.Property method calls (use TryGetEFPropertyArguments method), get set of IProperty. Ordering of equality comparisons is important.
- Once you get list of IProperties, match it with EntityType.FindPrimaryKey().Properties. 

If all above matches, then you have a Find query. We can relax the pattern bit more in future so user written query which are similar to that can also be optimized, for example doing First/Single/Last or default with key values all could be optimized.

Now if the pattern is identified then you create a new ShapedQueryExpression like below
```C#
            return new ShapedQueryExpression(
                selectExpression,
                new EntityShaperExpression(
                    entityType,
                    new ProjectionBindingExpression(
                        selectExpression,
                        new ProjectionMember(),
                        typeof(ValueBuffer)),
                    false));
````
But instead of using `selectExpression`, you need to add another custom expression ReadItemExpression. ReadItemExpression should contain properties/field for things you need to remember to run ReadItemAsync on Cosmos.
From comparisons above, you will have parameter values for "id" & Partition key. you store them on ReadItemExpression. And you return shaped query created above directly skipping whole translation pipeline. Also set ShapedQueryExpression's ResultCardinality here.

- Cosmos does not have QueryTranslationPostprocessor so no changes needed there.

Step 2: Taking care of materialization
In method `CosmosShapedQueryCompilingExpressionVisitor.VisitShapedQueryExpression`, 
You can check if shapedQueryExpression.QueryExpression is SelectExpression or ReadItemExpression.
if it is ReadItemExpression then you need to skip processing does for SelectExpression.
Hence
- ApplyProjection call is not needed.
- You would need to inject JObject & EntityMaterializers.
- The call to CosmosProjectionBindingRemovingExpressionVisitor.Visit is the case where you would need to write a different visitor which would bind the entityMaterializer to object being returned from ReadItemAsync method call. You can inspect expression tree at this point. EntityMaterializer takes a value buffer and generates entity instance. You would need to replace the calls to each value buffer read from particular index to the JObject you get from server side and property read. You can some inspiration from CosmosProjectionBindingRemovingExpressionVisitor. Especially you will need CreateGetValueExpression from that class.

Step 3 execute the query.
- Given the arguments for QueryingEnumerable, I would suggest using a different enumerable to be used with ReadItemAsync.
- You will get values of parameters from QueryContext.ParameterValues, get values from there and pass to ReadItemAsync.
- For ReadItemAsync, create a new method in CosmosClientWrapper which does that and being called from above queryingEnumerable you created.

We can expand a lot more in future. This is minimum amount of basic stuff we need to add in first pass.
Let me know if you have questions or get stuck anywhere. Expression trees are hard 😉

1iveowl commented:
> * Predicate is binary expression with ExpressionType.AndAlso where binaryExpression.Left is ExpressionType.Equal and binaryExpression.Right is either ExpressionType.AndAlso with recursive structure or Expression.Equal.

On inspection I found that Left/Right should be swapped. That is `binaryExpression.Left` can be either `Equal` or `AndAlso`  while `binaryExpression.Right` should be `Equal` when pattern matching for `Find(..)` - i.e. the code will look something like this:

```csharp
public override Expression Visit(Expression expression)
{
    if (expression is MethodCallExpression methodCallExpression
        && methodCallExpression.Method.IsGenericMethod
        && methodCallExpression.Method.Name == nameof(Queryable.FirstOrDefault))
    {
        if (methodCallExpression.Arguments[0] is MethodCallExpression queryRootMethodCallExpression
            && methodCallExpression.Method.IsGenericMethod
            && queryRootMethodCallExpression.Method.Name == nameof(Queryable.Where))
        {
            if (queryRootMethodCallExpression.Arguments[0] is QueryRootExpression queryRootExpression)
            {
                _entityType = queryRootExpression.EntityType;
            }

            if (queryRootMethodCallExpression.Arguments[1] is UnaryExpression unaryExpression
                && unaryExpression.Operand is LambdaExpression lambdaExpression
                && lambdaExpression.Body is BinaryExpression binaryExpression
                && binaryExpression.NodeType == ExpressionType.AndAlso
                && binaryExpression.Right.NodeType == ExpressionType.Equal
                && (binaryExpression.Left.NodeType == ExpressionType.AndAlso
                    || binaryExpression.Left.NodeType == ExpressionType.Equal))
            {
                _isFind = !(_entityType is null);
            }
        }
    }

    return base.Visit(expression);
}
```

smitpatel commented:
You would need recursive method to identify structure of predicate since, it would have more nesting as number of PK properties increases. We can defer it for later.

smitpatel commented:
Also don't set `_find` and call base visitor. Short-circuit entire thing.
Create a new ShapedQueryExpression as I described above.

1iveowl commented:
> You would need recursive method to identify structure of predicate since, it would have more nesting as number of PK properties increases. We can defer it for later.

Thank you. The recursive part makes better sense to me. I'd misunderstood the guide on this part and I had a misunderstanding on how the ExpressionVisitor works. I thought that it would traverse the tree by itself. It makes much more sense to with the recursive call to `VisitBinary`.

smitpatel commented:
It would be actually separate method. Rather than base call. We cannot use ExpressionVisitor for that.
Have a look at this code which essentially does the same, (though it is right leaning tree so you may have to flip a little)
https://github.com/dotnet/efcore/blob/bb88898f5b88db612862d9195fce72de873f8684/src/EFCore.InMemory/Query/Internal/InMemoryQueryableMethodTranslatingExpressionVisitor.cs#L449-L490
It collects Expressions from left & right. You will get EF.Property calls on left & ParameterExpression on right.

1iveowl commented:
> * Ordering of equality comparisons is important.

@smitpatel Can you elaboate here. By this do you mean that the property names from the EntitiType should be matched to the property names of the query by index? Or?

smitpatel commented:
If predicate is 
`EF.Property(a, "id")` == p0 && EF.Property(a, "PartitionKey") == p1)`
then you need to also remember the order of "id" and "PartitionKey" so when you match them with PrimaryKey.Properties, they appear in exact same order.

1iveowl commented:
@smitpatel I'm a bit stuck. Any hints on the following?

When looking at the return value from [`VisitShapedQueryExpression`]: https://github.com/dotnet/efcore/blob/cdc732f3ae7c60cdad78bcd6f0d22947ade90690/src/EFCore/Query/ShapedQueryCompilingExpressionVisitor.cs#L74

...I'm getting this exception:
> Extension node must override the property Expression.NodeType.

I have the sense that I might be missing something trivial, but my eyes can't catch it.

The new `VisitShapedQueryExpression` looks like this. 
```csharp
protected override Expression VisitShapedQueryExpression(ShapedQueryExpression shapedQueryExpression)
{
    Check.NotNull(shapedQueryExpression, nameof(shapedQueryExpression));

    var jObjectParameter = Expression.Parameter(typeof(JObject), "jObject");

    var shaperBody = shapedQueryExpression.ShaperExpression;
    shaperBody = new JObjectInjectingExpressionVisitor().Visit(shaperBody);
    shaperBody = InjectEntityMaterializers(shaperBody);

    switch (shapedQueryExpression.QueryExpression)
    {
        case SelectExpression selectExpression:
            selectExpression.ApplyProjection();

            shaperBody = new CosmosProjectionBindingRemovingExpressionVisitor(selectExpression, jObjectParameter, IsTracking)
                .Visit(shaperBody);

            var shaperLambda = Expression.Lambda(
                shaperBody,
                QueryCompilationContext.QueryContextParameter,
                jObjectParameter);

            return Expression.New(
                typeof(QueryingEnumerable<>).MakeGenericType(shaperLambda.ReturnType).GetConstructors()[0],
                Expression.Convert(
                    QueryCompilationContext.QueryContextParameter,
                    typeof(QueryContext)),
                Expression.Constant(_sqlExpressionFactory),
                Expression.Constant(_querySqlGeneratorFactory),
                Expression.Constant(selectExpression),
                Expression.Constant(shaperLambda.Compile()),
                Expression.Constant(_contextType),
                Expression.Constant(_partitionKeyFromExtension, typeof(string)),
                Expression.Constant(_logger));

        case ReadItemExpression readItemExpression:

            shaperBody =
                new CosmosProjectionBindingRemovingReadItemExpressionVisitor(readItemExpression, jObjectParameter, IsTracking)
                .Visit(shaperBody);

            var shaperReadItemLambda = Expression.Lambda(
                shaperBody,
                QueryCompilationContext.QueryContextParameter,
                jObjectParameter);

            return Expression.New(
                typeof(ReadItemQueryingEnumerable<>).MakeGenericType(shaperReadItemLambda.ReturnType).GetConstructors()[0],
                Expression.Convert(
                    QueryCompilationContext.QueryContextParameter,
                    typeof(QueryContext)),
                Expression.Constant(readItemExpression),
                Expression.Constant(shaperReadItemLambda.Compile()),
                Expression.Constant(_contextType),
                Expression.Constant(_logger));
            
        default:
            return null;
    }
}
```

The query expression that goes into `return Expression.New(typeof(ReadItemQueryingEnumerable<>)...)` looks like the following:

```csharp
.Lambda #Lambda1<System.Func`3[Microsoft.EntityFrameworkCore.Query.QueryContext,Newtonsoft.Json.Linq.JObject,Microsoft.EntityFrameworkCore.Cosmos.EndToEndCosmosTest+Customer_WithResourceId]>(
    Microsoft.EntityFrameworkCore.Query.QueryContext $queryContext,
    Newtonsoft.Json.Linq.JObject $jObject) {
    .Block(Newtonsoft.Json.Linq.JObject $jObject1) {
        $jObject1 = .Call Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosShapedQueryCompilingExpressionVisitor+CosmosProjectionBindingRemovingReadItemExpressionVisitor.SafeToObject(.Call $jObject.get_Item("c")
        );
        .If ($jObject1 == null) {
            null
        } .Else {
            .Block(
                Microsoft.EntityFrameworkCore.Storage.MaterializationContext $materializationContext1,
                Microsoft.EntityFrameworkCore.Metadata.IEntityType $entityType1,
                Microsoft.EntityFrameworkCore.Cosmos.EndToEndCosmosTest+Customer_WithResourceId $instance1,
                Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry $entry1,
                System.Boolean $hasNullKey1) {
                $materializationContext1 = .New Microsoft.EntityFrameworkCore.Storage.MaterializationContext(
                    .Constant<Microsoft.EntityFrameworkCore.Storage.ValueBuffer>(Microsoft.EntityFrameworkCore.Storage.ValueBuffer),
                    $queryContext.Context);
                $instance1 = null;
                $entry1 = .Call ($queryContext.StateManager).TryGetEntry(
                    .Constant<Microsoft.EntityFrameworkCore.Metadata.Internal.Key>(Key: Customer_WithResourceId.PartitionKey, Customer_WithResourceId.id, Customer_WithResourceId.Id PK),
                    .NewArray System.Object[] {
                        .Invoke (.Lambda #Lambda2<System.Func`2[Newtonsoft.Json.Linq.JToken,System.Object]>)(.Call $jObject1.get_Item("PartitionKey")
                        ),
                        (System.Object).Call Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosShapedQueryCompilingExpressionVisitor+CosmosProjectionBindingRemovingReadItemExpressionVisitor.SafeToObject(.Call $jObject1.get_Item("id")
                        ),
                        (System.Object).Call Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosShapedQueryCompilingExpressionVisitor+CosmosProjectionBindingRemovingReadItemExpressionVisitor.SafeToObject(.Call $jObject1.get_Item("Id")
                        )
                    },
                    True,
                    $hasNullKey1);
                .If (
                    !$hasNullKey1
                ) {
                    .If ($entry1 != null) {
                        .Block() {
                            $entityType1 = $entry1.EntityType;
                            $instance1 = (Microsoft.EntityFrameworkCore.Cosmos.EndToEndCosmosTest+Customer_WithResourceId)$entry1.Entity
                        }
                    } .Else {
                        .Block(Microsoft.EntityFrameworkCore.Storage.ValueBuffer $shadowValueBuffer1) {
                            $shadowValueBuffer1 = .Constant<Microsoft.EntityFrameworkCore.Storage.ValueBuffer>(Microsoft.EntityFrameworkCore.Storage.ValueBuffer);
                            $entityType1 = .Block(System.String $discriminator) {
                                $discriminator = .Call Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosShapedQueryCompilingExpressionVisitor+CosmosProjectionBindingRemovingReadItemExpressionVisitor.SafeToObject(.Call $jObject1.get_Item("Discriminator")
                                );
                                .Switch ($discriminator) {
                                .Case ("Customer_WithResourceId"):
                                        .Constant<Microsoft.EntityFrameworkCore.Metadata.IEntityType>(EntityType: Customer_WithResourceId)
                                .Default:
                                        .Block() {
                                            .Throw .Call Microsoft.EntityFrameworkCore.Query.EntityShaperExpression.CreateUnableToDiscriminateException(
                                                .Constant<Microsoft.EntityFrameworkCore.Metadata.Internal.EntityType>(EntityType: Customer_WithResourceId),
                                                (System.Object)$discriminator);
                                            null
                                        }
                                }
                            };
                            $instance1 = .Switch ($entityType1) {
                            .Case (.Constant<Microsoft.EntityFrameworkCore.Metadata.IEntityType>(EntityType: Customer_WithResourceId)):
                                    .Block() {
                                        $shadowValueBuffer1 = .New Microsoft.EntityFrameworkCore.Storage.ValueBuffer(.NewArray System.Object[] {
                                                (System.Object).Call Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosShapedQueryCompilingExpressionVisitor+CosmosProjectionBindingRemovingReadItemExpressionVisitor.SafeToObject(.Call $jObject1.get_Item("Discriminator")
                                                ),
                                                $jObject1
                                            });
                                        .Block(Microsoft.EntityFrameworkCore.Cosmos.EndToEndCosmosTest+Customer_WithResourceId $instance) {
                                            $instance = .New Microsoft.EntityFrameworkCore.Cosmos.EndToEndCosmosTest+Customer_WithResourceId();
                                            $instance.<PartitionKey>k__BackingField = .Invoke (.Lambda #Lambda3<System.Func`2[Newtonsoft.Json.Linq.JToken,System.Int32]>)(.Call $jObject1.get_Item("PartitionKey")
                                            );
                                            $instance.<id>k__BackingField = .Call Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosShapedQueryCompilingExpressionVisitor+CosmosProjectionBindingRemovingReadItemExpressionVisitor.SafeToObject(.Call $jObject1.get_Item("id")
                                            );
                                            $instance.<Id>k__BackingField = (System.Int32).Call Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosShapedQueryCompilingExpressionVisitor+CosmosProjectionBindingRemovingReadItemExpressionVisitor.SafeToObject(.Call $jObject1.get_Item("Id")
                                            );
                                            $instance.<Name>k__BackingField = .Call Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosShapedQueryCompilingExpressionVisitor+CosmosProjectionBindingRemovingReadItemExpressionVisitor.SafeToObject(.Call $jObject1.get_Item("Name")
                                            );
                                            $instance
                                        }
                                    }
                            .Default:
                                    null
                            };
                            $entry1 = .Call $queryContext.StartTracking(
                                $entityType1,
                                $instance1,
                                $shadowValueBuffer1);
                            $instance1
                        }
                    }
                } .Else {
                    .Default(System.Void)
                };
                $instance1
            }
        }
    }
}

.Lambda #Lambda2<System.Func`2[Newtonsoft.Json.Linq.JToken,System.Object]>(Newtonsoft.Json.Linq.JToken $var1) {
    .If (
        $var1 == .Default(Newtonsoft.Json.Linq.JToken) || $var1.Type == .Constant<Newtonsoft.Json.Linq.JTokenType>(Null)
    ) {
        .Default(System.Object)
    } .Else {
        (System.Object).Block(System.Int32 $parsed) {
            .If (
                .Call System.Int32.TryParse(
                    .Call $var1.ToObject(),
                    .Constant<System.Globalization.NumberStyles>(Any),
                    .Constant<System.IFormatProvider>(),
                    $parsed)
            ) {
                $parsed
            } .Else {
                0
            }
        }
    }
}

.Lambda #Lambda3<System.Func`2[Newtonsoft.Json.Linq.JToken,System.Int32]>(Newtonsoft.Json.Linq.JToken $var2) {
    .If (
        $var2 == .Default(Newtonsoft.Json.Linq.JToken) || $var2.Type == .Constant<Newtonsoft.Json.Linq.JTokenType>(Null)
    ) {
        .Default(System.Int32)
    } .Else {
        .Block(System.Int32 $parsed) {
            .If (
                .Call System.Int32.TryParse(
                    .Call $var2.ToObject(),
                    .Constant<System.Globalization.NumberStyles>(Any),
                    .Constant<System.IFormatProvider>(),
                    $parsed)
            ) {
                $parsed
            } .Else {
                0
            }
        }
    }
}
```

smitpatel commented:
> Extension node must override the property Expression.NodeType.

Make sure that you override `Type` & `NodeType` property in `ReadItemExpression` class.

1iveowl commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryContext.cs:
```diff
@@ -24,11 +26,15 @@ public class CosmosQueryContext : QueryContext
             [NotNull] QueryContextDependencies dependencies,
-            [NotNull] CosmosClientWrapper cosmosClient)
+            [NotNull] CosmosClientWrapper cosmosClient,
+            [NotNull] IValueGeneratorSelector valueGeneratorSector)
```
I need to the `ValueGeneratorSelector` to generate a resource id from the keys. 

I reasoned that this is part of the QueryContext, so injected it here.

1iveowl commented on src/EFCore.Cosmos/Diagnostics/Internal/CosmosLoggerExtensions.cs:
```diff
@@ -51,6 +52,28 @@ public static class CosmosLoggerExtensions
+                level => LoggerMessage.Define<string, string, string>(
+                    level,
+                    CoreEventId.ProviderBaseId,
+                    "Executing Read Item [Partition Key, Resource Id=[{parameters}]]{newLine}{commandText}"));
```
The right formating for logging Read Item?

AndriySvyryd replied:
This is fine for now

1iveowl commented on src/EFCore.Cosmos/Metadata/CosmosConstant.cs:
```diff
@@ -0,0 +1,16 @@
+        /// <summary>
+        ///     Cosmos DB resource id identifier: <c>id</c>.
+        /// </summary>
+        public const string ResourceIdIdentifier = "id";
```
Adding this as a context makes sense to me. Agree?

AndriySvyryd replied:
We already have it in `StoreKeyConvention.IdPropertyName`

1iveowl commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.CosmosProjectionBindingRemovingExpressionVisitor.cs:
```diff
@@ -2,740 +2,39 @@
     public partial class CosmosShapedQueryCompilingExpressionVisitor
     {
-        private sealed class CosmosProjectionBindingRemovingExpressionVisitor : ExpressionVisitor
+        private sealed class CosmosProjectionBindingRemovingExpressionVisitor : CosmosProjectionBindingRemovingExpressionVisitorBase
```
Moved the `CosmosProjectionBindingRemovingExpressionVisitor` into a base class, as the shaper is almost identical in the` SelectExpression` and `ReadItemExpression` scenario.

1iveowl commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.QueryingEnumerable.cs:
```diff
@@ -49,8 +50,10 @@ private sealed class QueryingEnumerable<T> : IEnumerable<T>, IAsyncEnumerable<T>
                 _logger = logger;
+
+                // TODO try also to extract partition key from query;
+                _partitionKey = partitionKeyFromExtension;
```
Would be nice to be able to extract the Partition key from the query too for `SelectExpression` scenario too.

AndriySvyryd replied:
That's tracked by https://github.com/dotnet/efcore/issues/20350

1iveowl commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.cs:
```diff
@@ -55,32 +55,63 @@ protected override Expression VisitShapedQueryExpression(ShapedQueryExpression s
+                        Expression.Constant(_logger));
+                    
+                default:
+                    return null;
```
Maybe throwing an exception in case the Expression is not known is better here?

smitpatel replied:
throw InvalidOperationException.

1iveowl commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,341 @@
+                    return TryGenerateResourceIdFromKeys(out resourceId);
+                }
+
+                private bool TryGenerateResourceIdFromKeys(out string resourceId)
```
Curious on your feedback on this approach?

1iveowl commented on test/EFCore.Cosmos.FunctionalTests/ValueGenerator/CustomPartitionKeyIdGenerator.cs:
```diff
@@ -0,0 +1,106 @@
+            }
+            
+            var partitionKey = entityType.GetPartitionKeyPropertyName() ?? CosmosClientWrapper.DefaultPartitionKey;
+            foreach (var property in pk.Properties.Where(p => p.Name != CosmosConstant.ResourceIdIdentifier))
```
Example of benefit of defining the `ResourceIdIdentifier` constant globally.

AndriySvyryd commented on src/EFCore.Cosmos/Diagnostics/Internal/CosmosLoggerExtensions.cs:
```diff
@@ -51,6 +52,28 @@ public static class CosmosLoggerExtensions
+                level => LoggerMessage.Define<string, string, string>(
+                    level,
+                    CoreEventId.ProviderBaseId,
+                    "Executing Read Item [Partition Key, Resource Id=[{parameters}]]{newLine}{commandText}"));
```
This is fine

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,314 @@
+                    
+                    var resourceIdProperty = _entityType.FindProperty(CosmosConstant.ResourceIdIdentifier);
+
+                    var resourceValueGenerator = CosmosQueryContext.ValueGeneratorSelector.Select(resourceIdProperty, _entityType);
```
Instead of this you can set the entry state to `Added`, read the value, then change it back to `Detached`

1iveowl replied:
OK.

smitpatel commented on src/EFCore.Cosmos/Extensions/CosmosServiceCollectionExtensions.cs:
```diff
@@ -63,6 +64,7 @@ public static IServiceCollection AddEntityFrameworkCosmos([NotNull] this IServic
                 // New Query pipeline
                 .TryAdd<IQueryableMethodTranslatingExpressionVisitorFactory, CosmosQueryableMethodTranslatingExpressionVisitorFactory>()
                 .TryAdd<IShapedQueryCompilingExpressionVisitorFactory, CosmosShapedQueryCompilingExpressionVisitorFactory>()
+                .TryAdd<IQueryCompiler, QueryCompiler>()
```
Revert changes in this file.

smitpatel commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.ReadItem.cs:
```diff
@@ -0,0 +1,147 @@
+{
+    public partial class CosmosQueryableMethodTranslatingExpressionVisitor
+    {
+        private IEntityType _entityType;
```
Remove private fields. Just use variable in appropriate part of code.

smitpatel commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.ReadItem.cs:
```diff
@@ -0,0 +1,147 @@
+                    if (queryRootMethodCallExpression.Arguments[0] is QueryRootExpression queryRootExpression)
+                    {
+                        _entityType = queryRootExpression.EntityType;
+                    }
```
if Arguments[0] is not QueryRootExpression then it should short-circuit rather than processing Arguments[1]

smitpatel commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.ReadItem.cs:
```diff
@@ -0,0 +1,147 @@
+                        && unaryExpression.Operand is LambdaExpression lambdaExpression
+                        && lambdaExpression.Body is BinaryExpression binaryExpression)
+                    {
+                        var (left, right) = DecomposeJoinCondition(binaryExpression);
```
Call ProcessJoinCondition passing List<Expression> for left & right, so we can avoid the whole DecomposeArrayExpression. Remove all other irrelevant code.

smitpatel commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.ReadItem.cs:
```diff
@@ -0,0 +1,147 @@
+        /// <summary>Dispatches the expression to one of the more specialized visit methods in this class.</summary>
+        /// <param name="expression">The expression to visit.</param>
+        /// <returns>The modified expression, if it or any subexpression was modified; otherwise, returns the original expression.</returns>
+        public override Expression Visit(Expression expression)
```
Put in same file as other code. No need to create partial class.

1iveowl replied:
Should we combine with new base class too? I.e. all three files in one file?

smitpatel replied:
There is no new base class here. Only 2 files for this class.

smitpatel commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,318 @@
+
+            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+            public string ToQueryString()
```
@AndriySvyryd - We need to improve this.

AndriySvyryd replied:
File an issue

smitpatel commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,318 @@
+                /// <summary>Advances the enumerator to the next element of the collection.</summary>
+                /// <returns>
+                /// <see langword="true" /> if the enumerator was successfully advanced to the next element; <see langword="false" /> if the enumerator has passed the end of the collection.</returns>
+                /// <exception cref="T:System.InvalidOperationException">The collection was modified after the enumerator was created.</exception>
```
Remove XML comments.

smitpatel commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,318 @@
+                }
+            }
+
+            private abstract class EnumerableBase
```
ReadItemEnumerableBase

smitpatel commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,318 @@
+                        return rid;
+                    }
+                    
+                    object CreateEntityInstanceWithPrimaryKeyValues()
```
@AndriySvyryd - Is there any alternative here?

AndriySvyryd replied:
Not that I can think of, perhaps @ajcvickers has an idea

smitpatel commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitorFactory.cs:
```diff
@@ -49,9 +49,11 @@ public virtual ShapedQueryCompilingExpressionVisitor Create(QueryCompilationCont
         {
             Check.NotNull(queryCompilationContext, nameof(queryCompilationContext));
 
+            var cosmosQueryCompilationContext = (CosmosQueryCompilationContext)queryCompilationContext;
```
revert

smitpatel commented on src/EFCore.Cosmos/Query/Internal/ReadItemExpression.cs:
```diff
@@ -0,0 +1,99 @@
+    /// </summary>
+    public class ReadItemExpression : Expression
+    {
+
```
Remove blank line.

smitpatel commented on src/EFCore.Cosmos/Query/Internal/ReadItemExpression.cs:
```diff
@@ -0,0 +1,99 @@
+
+        /// <summary>Gets the static type of the expression that this <see cref="T:System.Linq.Expressions.Expression" /> represents.</summary>
+        /// <returns>The <see cref="T:System.Type" /> that represents the static type of the expression.</returns>
+        public override Type Type => typeof(JObject);
```
Copy same this is internal API XML docs here.

smitpatel replied:
We should make this Enumerable of JObject from query consistency perspective.

smitpatel commented on src/EFCore.Cosmos/Query/Internal/ReadItemExpression.cs:
```diff
@@ -0,0 +1,99 @@
+
+        /// <summary>Gets the node type of this <see cref="T:System.Linq.Expressions.Expression" />.</summary>
+        /// <returns>One of the <see cref="T:System.Linq.Expressions.ExpressionType" /> values.</returns>
+        public override ExpressionType NodeType => ExpressionType.Extension;
```
Same as above.

smitpatel commented on src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs:
```diff
@@ -496,7 +496,27 @@ private static void ProcessResponse(ResponseMessage response, IUpdateEntry entry
+        {
+            _commandLogger.ExecutingReadItem(partitionKey, resourceId);
+
+            return new DocumentEnumerable(this, containerId, partitionKey, resourceId, true, null);
```
untangle this. ReadItem does not return an enumerable. It should not use DocumentEnumerable.

smitpatel commented on test/EFCore.Cosmos.FunctionalTests/ValueGenerator/CustomPartitionKeyIdGenerator.cs:
```diff
@@ -0,0 +1,105 @@
+
+namespace Microsoft.EntityFrameworkCore.Cosmos.ValueGenerator
+{
+    public class CustomPartitionKeyIdGenerator : EntityFrameworkCore.ValueGeneration.ValueGenerator
```
Add using and simplify type usage.

smitpatel commented on test/EFCore.Cosmos.FunctionalTests/ValueGenerator/CustomPartitionKeyIdValueGeneratorFactory.cs:
```diff
@@ -0,0 +1,21 @@
+        /// </summary>
+        /// <param name="property"> The property to create the value generator for. </param>
+        /// <returns> The newly created value generator. </returns>
+        public override EntityFrameworkCore.ValueGeneration.ValueGenerator Create(IProperty property)
```
No XML docs in tests.

AndriySvyryd replied:
Add `using EntityFrameworkCore.ValueGeneration;`

1iveowl replied:
> Add `using EntityFrameworkCore.ValueGeneration;`

Can only do that if we change the `namespace`.

AndriySvyryd replied:
Yes, this should be in `TestUtilities`

AndriySvyryd replied:
Don't forget to move this

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -279,34 +276,41 @@ protected bool TryGetResourceId(out string resourceId)
 
-                    var resourceValueGenerator = CosmosQueryContext.ValueGeneratorSelector.Select(resourceIdProperty, _entityType);
+                        var entityProperties = entityEntry.GetType().GetProperties()
+                            .Where(p => p.Name != StoreKeyConvention.IdPropertyName);
```
This filter is redundant

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -279,34 +276,41 @@ protected bool TryGetResourceId(out string resourceId)
-                    resourceId = resourceValueGenerator.Next(entry).ToString();
+                        foreach (var entityProperty in entityProperties)
+                        {
+                            if (_propertyValueDictionary.ContainsKey(entityProperty.Name))
```
Use `TryGet` instead

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,318 @@
+                    {
+                        resourceId = value.ToString();
+
+                        if (!string.IsNullOrEmpty(resourceId))
```
@1iveowl When would `resourceId` be null or empty? Should we throw instead?

1iveowl replied:
If no `resourceId` is specified as `null/string.Empty` part of the `Find()`, then we try to generate it, from the other primary keys. Thus I don't think we should necessary throw here. 

If we are unable to get or generate the `ResourceId`, then we already throw in the MoveNext section.

AndriySvyryd replied:
If "id" is part of the PK then a valid value needs to be supplied in `Find`. If it's not then it wouldn't be in `_propertyValueDictionary`

1iveowl commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -27,6 +28,7 @@ public class CosmosQueryableMethodTranslatingExpressionVisitor : QueryableMethod
         private readonly ISqlExpressionFactory _sqlExpressionFactory;
         private readonly CosmosSqlTranslatingExpressionVisitor _sqlTranslator;
         private readonly CosmosProjectionBindingExpressionVisitor _projectionBindingExpressionVisitor;
+        private IEnumerable<(IProperty, string)> _propertyParameterList;
```
Move to local

1iveowl commented:
@dougbu you can ignore this request for review. Was having some challenges with git rebasing, which must have triggered the request. Thx.

1iveowl commented:
@smitpatel and @AndriySvyryd I've cleaned up the code, rebased to the latest master. I've added throwing of a few exceptions according to the review, not sure if the exceptions kind I've picked align with the conventions you are adhering too and likewise with the exception text. Let me me know if it doesn't. Thx.

smitpatel commented on src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs:
```diff
@@ -519,126 +523,129 @@ private static void ProcessResponse(ResponseMessage response, IUpdateEntry entry
             IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
 
-            private sealed class Enumerator : IEnumerator<JObject>
+            private sealed class Enumerator : EnumerableBase, IEnumerator<JObject>
```
Revert all the changes related to this.

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -68,6 +69,114 @@ public class CosmosQueryableMethodTranslatingExpressionVisitor : QueryableMethod
+                                            .Select(UnwrapParameter),
+                                        (property, parameter) => (property, parameter));
+
+                                    var sharpedQueryExpression = new ShapedQueryExpression(
```
```suggestion
                                    var shapedQueryExpression = new ShapedQueryExpression(
```

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,310 @@
+                                }
+                                else
+                                {
+                                    throw new ArgumentException("Partition key missing and/or cosmos resource id is missing or cannot be generated");
```
throw InvalidOperationException

1iveowl replied:
> use CosmosStrings

Please add more context?

AndriySvyryd replied:
> Please add more context?

Add the message as a resource to `CosmosStrings.resx` then run `CosmosStrings.Designer.tt`

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,310 @@
+                    
+                    if (CosmosQueryContext.ParameterValues.TryGetValue(_readItemExpression.PartitionKeyParameterName, out var partitionKeyObj))
+                    {
+                        partitionKey = partitionKeyObj.ToString();
```
cast to string instead

1iveowl replied:
In the existing tests - i.e. before extending with `Find()` - `PartitionKey` have been defined as an `int`, hence this cast will fail unless we decided that `PartitionKey` can from now on only be defined as a `string`?

1iveowl replied:
OK. This introduces the requirement that ParitionKey is defined as `string` or that the property has a converter like: `cb.Property(c => c.PartitionKey).HasConversion<string>()`

AndriySvyryd replied:
We have two options: either we preserve the original type and use the appropriate constructor of `PartitionKey`  or we require it to be string or have a converter

1iveowl replied:
> We have two options: either we preserve the original type and use the appropriate constructor of `PartitionKey` or we require it to be string or have a converter

I think it makes sense to require a string, since Cosmos ultimately stores the partition key as a string anyhow.

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,310 @@
+                {
+                    if (_propertyValueDictionary.TryGetValue(StoreKeyConvention.IdPropertyName, out var value))
+                    {
+                        resourceId = value.ToString();
```
Use `as` and add to the exception message that id needs to have a string value

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,310 @@
+                        var internalEntityEntry = new InternalEntityEntryFactory().Create(_stateManager, _entityType, entityEntry);
+                        var entry = new EntityEntry(internalEntityEntry) { State = EntityState.Added };
+
+                        var rid = entry.Members?
```
Yes, we'll validate it in https://github.com/dotnet/efcore/issues/17751

smitpatel replied:
Can use `entry[StoreKeyConvention.IdPropertyName]`

AndriySvyryd replied:
No, I wasn't exact. There will always be a member with `id` as the store property name. So the condition should be on `member.Metadata.GetJsonPropertyName()`

1iveowl replied:
`member.Metadata.GetJsonPropertyName()` ?

AndriySvyryd replied:
We are looking for the property mapped to `id`:
`.FirstOrDefault(member => member.Metadata.GetJsonPropertyName() == StoreKeyConvention.IdPropertyName)`

1iveowl replied:
Maybe I'm missing a `using`, `member.Metadata.GetJsonPropertyName` cannot be resolved.

AndriySvyryd replied:
Oh, it's on `IProperty, use `entry.Properties` instead

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.cs:
```diff
@@ -55,32 +55,63 @@ protected override Expression VisitShapedQuery(ShapedQueryExpression shapedQuery
+                        Expression.Constant(_logger));
+                    
+                default:
+                    throw new InvalidOperationException();
```
Add a message or throw a NotImplementedException

AndriySvyryd commented on src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs:
```diff
@@ -499,6 +499,61 @@ private static void ProcessResponse(ResponseMessage response, IUpdateEntry entry
+
+        private static JObject JObjectFromReadItemResponseMessage(ResponseMessage responseMessage)
+        {
+            if (responseMessage.IsSuccessStatusCode)
```
call `response.EnsureSuccessStatusCode()`

AndriySvyryd commented on src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs:
```diff
@@ -499,6 +499,61 @@ private static void ProcessResponse(ResponseMessage response, IUpdateEntry entry
+                var reader = new StreamReader(responseStream);
+                var jsonReader = new JsonTextReader(reader);
+
+                var jObj = new JsonSerializer().Deserialize<JObject>(jsonReader);
```
```suggestion
                var jObject = new JsonSerializer().Deserialize<JObject>(jsonReader);
```

AndriySvyryd commented on src/EFCore.Cosmos/ValueGeneration/Internal/IdValueGenerator.cs:
```diff
@@ -47,7 +48,7 @@ protected override object NextValue(EntityEntry entry)
 
             var partitionKey = entityType.GetPartitionKeyPropertyName() ?? CosmosClientWrapper.DefaultPartitionKey;
-            foreach (var property in pk.Properties)
+            foreach (var property in pk.Properties.Where(p => p.Name != StoreKeyConvention.IdPropertyName))
```
Undo this. If `id` is part of the PK this generator can't be used

1iveowl replied:
OK, but why iterate over the value we are generating, rather than just filtering it out?

smitpatel commented on src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs:
```diff
@@ -545,100 +613,63 @@ private sealed class DocumentEnumerable : IEnumerable<JObject>
-                    _reader = null;
-                    _responseStream.Dispose();
-                    _responseStream = null;
+                    Dispose();
```
This will null out _responseMessage also which is incorrect.

smitpatel commented on src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs:
```diff
@@ -749,15 +749,94 @@ public async ValueTask<bool> MoveNextAsync()
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private void SetCurrent()
```
Remove method and inline in usage.

smitpatel commented on src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs:
```diff
@@ -749,15 +749,94 @@ public async ValueTask<bool> MoveNextAsync()
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            protected void SetCurrentToDefault()
```
Inline.

1iveowl commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -68,6 +69,114 @@ public class CosmosQueryableMethodTranslatingExpressionVisitor : QueryableMethod
+            {
+                if (methodCallExpression.Arguments[0] is MethodCallExpression queryRootMethodCallExpression
+                    && methodCallExpression.Method.IsGenericMethod
+                    && queryRootMethodCallExpression.Method.Name == nameof(Queryable.Where))
```
Changed to `QueryableMethods.FirstOrDefaultWithPredicate`

1iveowl replied:
`FirstOrDefault` changed to `QueryableMethods.FirstOrDefaultWithoutPredicate`.

1iveowl commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -68,6 +69,114 @@ public class CosmosQueryableMethodTranslatingExpressionVisitor : QueryableMethod
+                                    return sharpedQueryExpression;
+                                }
+
+                                static Expression UnwrapConvertExpression(Expression expr)
```
Can't use `expression` as this variable is already defined - i.e.: `public override Expression Visit(Expression expression)`

I've used `wrappedExpression`. Other options are:
- inline the function. 
- change the function from local to private for the class.

1iveowl commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,310 @@
+                                    throw new ArgumentException("Partition key missing and/or cosmos resource id is missing or cannot be generated");
+                                }
+
+                                var hasNext = !(_item is null);
```
Now made explicit/remembered with `_hasExecuted` boolean.

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -68,6 +69,97 @@ public class CosmosQueryableMethodTranslatingExpressionVisitor : QueryableMethod
+                            {
+                                var entityTypePrimaryKeyProperties = entityType.FindPrimaryKey().Properties;
+
+                                if (queryProperties.SequenceEqual(entityTypePrimaryKeyProperties))
```
Also check that if there's a partition key it must be part of the PK.

Add a test with an entity type that has the partition key that's not part of the PK.
Add a test with an entity type that doesn't have a partition key.

1iveowl replied:
> Also check that if there's a partition key it must be part of the PK.

There has to be a partition key for `Find()` to work, right? I can't find anything official, but according to this SO post, non-partitioned containers are being deprecated: https://stackoverflow.com/a/55138587/4140832.

1iveowl replied:
> Add a test with an entity type that has the partition key that's not part of the PK.
> Add a test with an entity type that doesn't have a partition key.

Are the two not identical, since it is not possible to add the partition key to the PK, id the partition key is not defined for the entity?

AndriySvyryd replied:
The partition key is optional in EF, in that case we use the default one.
If there's no explicit partition key or if it's not part of the PK Find **should still work**, but without this optimization. I'll file an issue to log a warning in these cases.

AndriySvyryd commented on src/EFCore.Cosmos/Properties/CosmosStrings.resx:
```diff
@@ -168,4 +168,7 @@
     <value>Reverse is not supported without Limit or Offset.</value>
   </data>
+  <data name="InvalidResourceId" xml:space="preserve">
+    <value>Invalid Resource id. Resource id cannot be null or empty and must be a string value.</value>
```
```suggestion
    <value>Invalid id value specified in 'Find' call. The id value must be a string and it cannot be null or empty.</value>
```

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,307 @@
+                {
+                    partitionKey = null;
+
+                    var partitionKeyProperty = _entityType.FindProperty(_entityType.GetPartitionKeyPropertyName());
```
If there's no partition key property use `PartitionKey.None` in the `ReadItemAsync` call

1iveowl replied:
> If there's no partition key property use `PartitionKey.None` in the `ReadItemAsync` call

I guess this is also the way we should do it for `FeedIterator CreateQuery`?

1iveowl replied:
Should we handle the [`PartitionKey.Null`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.partitionkey.null?view=azure-dotnet) case also?

AndriySvyryd replied:
No, EF Core doesn't allow nulls in keys, even though Cosmos might.

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -162,6 +157,21 @@ public override Expression Visit(Expression expression)
+                                {
+                                    partitionKeyProperty = entityType.FindProperty(entityType.GetPartitionKeyPropertyName());
+                                }
+                                catch (ArgumentNullException)
```
Check for null instead if catching

AndriySvyryd commented on src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs:
```diff
@@ -575,15 +576,16 @@ private static JObject JObjectFromReadItemResponseMessage(ResponseMessage respon
         private async Task<ResponseMessage> CreateSingleItemQuery(
             string containerId,
-            string partitionKey,
+            string partitionKeyStr,
```
The previous name was better

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,307 @@
+
+                protected bool TryGetResourceId(out string resourceId)
+                {
+                    var ressourceIdProperty = _entityType.FindProperty(StoreKeyConvention.IdPropertyName);
```
```suggestion
                    var resourceIdProperty = _entityType.FindProperty(StoreKeyConvention.IdPropertyName);
```

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,307 @@
+                                }
+                                else
+                                {
+                                    throw new InvalidOperationException("Partition key missing and/or cosmos resource id is missing or cannot be generated");
```
Separate into two messages and move them to CosmosStrings

AndriySvyryd commented on src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs:
```diff
@@ -519,7 +572,61 @@ private static void ProcessResponse(ResponseMessage response, IUpdateEntry entry
+                    {
+                        if (jsonReader.TokenType == JsonToken.StartArray)
+                        {
+                            goto ObjectFound;
```
You can just return jsonReader

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/ReadItemExpression.cs:
```diff
@@ -0,0 +1,100 @@
+        /// </summary>
+        public ReadItemExpression(
+            [NotNull] IEntityType entityType,
+            [NotNull] IEnumerable<(IProperty property, string parameter)> propertyParameters)
```
Change this to `Dictionary<IProperty, string>`

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,307 @@
+
+                            if (TryGetParameterValue(property, out var value))
+                            {
+                                entityProperty.SetValue(entityEntry, value);
```
Combine this method with `GetGeneratedResourceId` and use this instead:
```
internalEntityEntry[property] = value;
```

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,307 @@
+
+                        var rid = entry.Properties
+                            .FirstOrDefault(propertyEntry => propertyEntry.Metadata.GetJsonPropertyName() == StoreKeyConvention.IdPropertyName)
+                            .CurrentValue.ToString();
```
cast to string instead

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,307 @@
+
+                    if (TryGetParameterValue(partitionKeyProperty, out var value))
+                    {
+                        partitionKey = value.ToString();
```
cast to string

1iveowl replied:
This now throws if `value` is an `int`. Should we require that the partition key is string?

1iveowl replied:
`cb.Property(c => c.PartitionKey).HasConversion<string>()` makes no difference. Seems we are missing a converter somewhere?

AndriySvyryd replied:
```C#
            var converter = partitionKeyProperty.GetTypeMapping().Converter;
            if (converter != null)
            {
                partitionKey = converter.ConvertToProvider(value);
            } else {
                partitionKey = (string)value;
            }
```

And do the same for `id`

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -68,6 +69,114 @@ public class CosmosQueryableMethodTranslatingExpressionVisitor : QueryableMethod
+                            bool ProcessJoinCondition(
+                                Expression joinCondition, ICollection<IProperty> properties, ICollection<string> paramNames)
+                            {
+                                if (joinCondition is BinaryExpression binaryExpr)
```
Spell out Expression in all variables

1iveowl replied:
Already have a `binarryExpression` in scope. I can change the existing `binaryExpression` to `lambdaBodyBinaryExpression`: 

```csharp
if (queryRootMethodCallExpression.Arguments[1] is UnaryExpression unaryExpression
                            && unaryExpression.Operand is LambdaExpression lambdaExpression
                            && lambdaExpression.Body is BinaryExpression lambdaBodyBinaryExpression)
```

Is that better?

1iveowl commented:
> Looks good. Please squash and rebase on latest master

When rebasing to latest master, I no longer can open the solution. Not sure what's going on.

AndriySvyryd commented:
> When rebasing to latest master, I can no longer open the solution. Not sure what's going on. All projects end up unloaded.

Run `build.cmd` then open the solution using `startvs.cmd`

1iveowl commented:
> > When rebasing to latest master, I can no longer open the solution. Not sure what's going on. All projects end up unloaded.
> 
> Run `build.cmd` then open the solution using `startvs.cmd`

Ahhh, the framework got updated.

1iveowl commented:
I see you've added netcoreapp5.0 tests now too. I assume this means that the tests I've written and/or changed for 3.1 need to be updated for 5.0 too?

I can open the solution with `startvs.cmd` now. I get a number of build errors, like this now:

>  Microsoft.EntityFrameworkCore.ChangeTracking.EntityEntry is an internal API that supports the Entity Framework Core infrastructure and not subject to the same compatibility standards as public APIs. It may be changed or removed without notice in any release.	

I will continue looking at it tomorrow.

AndriySvyryd commented:
Suppress that warning in the source code

1iveowl commented:
> Suppress that warning in the source code

Done.

1iveowl commented:
Regarding *squashing the commits*, I cannot get rid of the tree rebase to master I've done, so unable to squash below 4 commits. 

Probably due to my git skills coming to a short here.

AndriySvyryd commented:
> Regarding squashing the commits, I cannot get rid of the 3 rebase to master I did along the way, so unable to squash below 3 commits.

Good enough. Thanks for your contribution!

smitpatel commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -68,6 +69,115 @@ public class CosmosQueryableMethodTranslatingExpressionVisitor : QueryableMethod
+                                                && ProcessJoinCondition(binaryExpression.Right, properties, paramNames);
+
+                                        case ExpressionType.Equal:
+                                            if (binaryExpression.Left is MethodCallExpression methodCallExpr
```
full Expression

1iveowl replied:
Good catch. I can create a new PR to fix this?

smitpatel replied:
Just include in #20690

smitpatel commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -68,6 +69,115 @@ public class CosmosQueryableMethodTranslatingExpressionVisitor : QueryableMethod
+                                            {
+                                                if (methodCallExpr.TryGetEFPropertyArguments(out _, out var propertyName))
+                                                {
+#pragma warning disable EF1001
```
Why is this here?

1iveowl replied:
After rebasing to latest master it was no longer possible to compile without it.

@AndriySvyryd want to comment?

AndriySvyryd replied:
I'll fix this one by making the extension method public, but in general you also need to add a corresponding `#pragma warning restore EF1001`

smitpatel commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:
```diff
@@ -0,0 +1,323 @@
+                    {
+                        var property = _entityType.FindProperty(entityProperty.Name);
+
+                        if (TryGetParameterValue(property, out var parameterValue))
```
@AndriySvyryd - Shouldn't this be enumerated over key properties rather than all properties (or may be just over properties stored in ReadItem.

AndriySvyryd replied:
I'll fix this one
